### PR TITLE
Create initial working version of API sync activity digest page

### PR DESCRIPTION
**Pull request by ChemiKyle**
This PR adds a new project-level page named "API Sync - Latest Activity":

image

Customer's request:

> the current logging page is pretty good but it’s not easy to figure out the latest status of each site. A table with each site and the datetime, status, and number of records pulled, in the last data pull.

I clarified via email that they indeed want info on the latest import for each url defined in the module configuration.

### CODE REVIEW

File: get-log-digest.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+$all_accounted = false;
+do  {
+	$results = $module->queryLogs("
+	select log_id, timestamp, message, failure, $hasDetailsClause as hasDetails
+	where external_module_id = ? and project_id = ?
+	and log_id >= ? and log_id <= ?
+	and $message_subtrings
+	order by log_id desc
+", [$this_module_id, $module->getProjectId(), $end_log_id, $start_log_id]);
+
```
Comment (ChemiKyle) [inline get-log-digest.php:46]: Please scrutinize this section as you will, I wasn't sure how to properly go about sanitization of what equates to `WHERE column LIKE '$foo%'` statement "preparation".

File: classes/DigestLog.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    private function resetStatus(): void {
+        $this->accumulated_record_imports = 0;
+        $this->error_log_id = null;
+
+        foreach ($this->status_flags as &$flag) { $flag = false; }
+    }
+
+
+    private function cumRecords(array $row): void {
```
[the remark this answers is not in the document]
Comment (ChemiKyle) [reply classes/DigestLog.php:52]: Hahaha, my intro to programming was using MATLAB for linear algebra, all variants of "accumulative" get this funny 3 letter abbreviation. I'm not a pervert, I'm just following scientific programming standards!

Receipts for my piety:
https://www.mathworks.com/help/matlab/ref/cumsum.html
https://numpy.org/doc/stable/reference/generated/numpy.cumsum.html

File: APISyncExternalModule.php [hunk 3]
```diff
@@ -1651,11 +1652,26 @@ function requireDateParameter($paramName, $dateFormat, $array = null){
...
+if (version_compare(PHP_VERSION, '8.0.0', "<")) {
+	function str_starts_with($haystack, $needle): bool {
+		// discussion for why strncmp was chosen: https://stackoverflow.com/a/2792233/7418735
+		return (strncmp($haystack, $needle, strlen($needle)) === 0);
+	}
+
+	function str_contains($haystack, $needle): bool {
+		return (strpos($haystack, $needle) !== false);
+	}
+}
```
[the remark this answers is not in the document]
Comment (ChemiKyle) [reply APISyncExternalModule.php:1677]: I just spun up a 7.4 container and disabled that block and indeed it works fine. I suppose my only concern is that some EM users might be running a version of REDCap that predates the polyfill. I could remove this block and just add a `min_version` to the config.

I see it's using this repo: https://github.com/symfony/polyfill-php80 do you happen to know the first version of REDCap that implemented this?

[the remark this answers is not in the document]
Comment (mmcev106) [reply APISyncExternalModule.php:1677]: Looks like it was added three years ago in `11.2.0`:
```
mark@localhost:~/sites/redcap/redcap_v14.3.10$ git log --oneline Libraries/vendor/symfony/polyfill-php80/Php80.php 
48505f1f50 12.2.7: Composer update
b5e4521e64 11.2.0: Change: PHP 7.2.5 is now the new minimum PHP version that is required for running REDCap + composer update
mark@localhost:~/sites/redcap/redcap_v14.3.10$ 
```

I would personally remove them (and the `array_key_first()` implementation above if you feel like), then set `redcap-version-min` to `11.2.0`.

File: classes/DigestLog.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmcev106) [reply classes/DigestLog.php:52]: That's what I figured 😂.  Most of our developers do not have a scientific programming background (including me).  It of course doesn't matter significantly in a single case, but in general I'd avoid using abbreviations (that aren't extremely generic like URL), as there's a good chance they won't have meaning to the next developer that works on the module.

**Review by mmcev106 — APPROVED** [review]
All the latest commits are approved!  The open conversations are minor, and shouldn't prevent deployment regardless.

File: classes/DigestLog.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChemiKyle) [reply classes/DigestLog.php]: I thought it was standard but I went ahead and renamed to improve clarity and avoid hilarity in [f6cdb4e](https://github.com/vanderbilt-redcap/api-sync-module/pull/25/commits/f6cdb4e30302a35277d387c85e2082876b667e10)

File: APISyncExternalModule.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChemiKyle) [reply APISyncExternalModule.php]: Shims removed and min redcap version mandated in [c5a4049](https://github.com/vanderbilt-redcap/api-sync-module/pull/25/commits/c5a40490c97b0d09aef6d5e68897b6f62e068caf)

**Comment by ChemiKyle** [comment]
I need to re-introduce time windowing by default on the log pages (sane default is probably 1 day), a process had to be killed trying to churn through production data.

File: classes/DigestLog.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmcev106) [reply classes/DigestLog.php]: I appreciate it, and suspect other future developers will too 😊.

File: APISyncExternalModule.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmcev106) [reply APISyncExternalModule.php]: Thank you!

**Pull request merged**